compiler/ilgen/IlValue.cpp

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2016, 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+
+
+#include "compile/Compilation.hpp"
+#include "il/Block.hpp"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/Symbol.hpp"
+#include "il/symbol/AutomaticSymbol.hpp"
+#include "il/SymbolReference.hpp"
+#include "ilgen/IlValue.hpp" // must follow include for compile/Compilation.hpp for TR_Memory
+#include "ilgen/MethodBuilder.hpp"
+
+
+OMR::IlValue::IlValue(TR::Node *node, TR::TreeTop *treeTop, TR::Block *block, TR::MethodBuilder *methodBuilder)
+   : _id(methodBuilder->getNextValueID()),
+     _nodeThatComputesValue(node),
+     _treeTopThatAnchorsValue(treeTop),
+     _blockThatComputesValue(block),
+     _methodBuilder(methodBuilder),
+     _symRefThatCanBeUsedInOtherBlocks(0)
+   {
+   }
+
+TR::DataType
+OMR::IlValue::getDataType()
+   {
+   return _nodeThatComputesValue->getDataType();
+   }
+
+void
+OMR::IlValue::storeToAuto()
+   {
+   if (_symRefThatCanBeUsedInOtherBlocks == NULL)
+      {
+      TR::Compilation *comp = TR::comp();
+
+      // first use from another block, need to create symref and insert store tree where node  was computed
+      TR::SymbolReference *symRef = comp->getSymRefTab()->createTemporary(_methodBuilder->methodSymbol(), _nodeThatComputesValue->getDataType());
+      symRef->getSymbol()->setNotCollected();
+      char *name = (char *) comp->trMemory()->allocateHeapMemory(8 * sizeof(char));
+      sprintf(name, "_T%-5d", symRef->getCPIndex());
+      symRef->getSymbol()->getAutoSymbol()->setName(name);
+
+      // create store and its treetop
+      TR::Node *storeNode = TR::Node::createStore(symRef, _nodeThatComputesValue);
+      TR::TreeTop *prevTreeTop = _treeTopThatAnchorsValue->getPrevTreeTop();
+      TR::TreeTop *newTree = TR::TreeTop::create(comp, storeNode);
+      newTree->insertNewTreeTop(prevTreeTop, _treeTopThatAnchorsValue);
```

Comment:
Actually I was thinking that any future service that wants to use some kind of CHK opcode would explicitly generate the check as a treetop (there are some examples in IlBuilder where genTreeTop() is called directly, without using newValue() ).

The value to be checked should be created using newValue(), which would append *another* treetop with the same node as its child (even though this treetop is redundant in this case because the check treetop will have already anchored the node). If that value were ever used outside the block containing the node, then the another treetop (wow, sounds like a patent application now) would be replaced with a store. 

Imagine a DIVCHK is created for a new value that is the result of an idiv.

```
DIVCHK
    idiv
       left
       right
treetop
    ==>idiv
```

If the IlValue associated with the idiv node is used outside the block, the existing code in IlValue will transform that second treetop so that it looks like:
```
DIVCHK
    idiv
       left
       right
istore <temp slot X>
    ==>idiv
```

So the DIVCHK, because it precedes the treetop that newValue adds, is the *real* anchor point for the idiv. When there is no CHK opcode, the treetop inserted by newValue guarantees the right anchor point.

Dead trees is going to be (PR coming :) ) the first optimization to run, and it will end up removing lots of these treetops if they aren't really needed as anchor points.